File: strings/0028-strStr.cpp
/*
 * leetcode 28
 * find the index of the first occurence in a string
 *
 */

// brute force
// find needle in haystack
//
class Solution {
public:
    int strStr(string haystack, string needle) {
        int n = haystack.size(), m = needle.size();
        for (int i = 0; i + m <= n; i++) {
            bool flag = true;
            for (int j = 0; j < m; j++) {
                if (haystack[i + j] != needle[j]) {
                    flag = false;
                    break;
                }
            }
            if (flag) {
                return i;
            }
        }
        return -1;
    }
};

//
// KMP
//
/*

字符串的前缀是指不包含最后一个字符的所有以第一个字符开头的连续子串； 
字符串的后缀是指不包含第一个字符的所有以最后一个字符结尾的连续子串。

next数组存放的是当前长度下的最长相同前后缀的长度

 */

File: strings/0028-strStr.py
"""
leetcode 28 strStr
find a string needle in another string haystack
"""

# 
# brute force
#
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        L, n = len(needle), len(haystack)

        for start in range(n - L + 1):
            if haystack[start:start + L] == needle:
                return start
        return -1

#
# use the find function
#
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        return haystack.find(needle)
